allennlp/data/dataset_readers/dataset_reader.py

```diff
@@ -1,20 +1,53 @@
-from allennlp.data.dataset import InstanceCollection
+from typing import Iterator
+
+from allennlp.data.dataset import Dataset, InstanceCollection, LazyDataset
 from allennlp.data.instance import Instance
 from allennlp.common import Params
 from allennlp.common.registrable import Registrable
 
 
 class DatasetReader(Registrable):
     """
-    A ``DatasetReader`` reads data from some location and constructs a :class:`Dataset`.  All
-    parameters necessary to read the data apart from the filepath should be passed to the
-    constructor of the ``DatasetReader``.
+    A ``DatasetReader`` reads data from some location and constructs an
+    :class:`InstanceCollection`.  All parameters necessary to read the data apart from the filepath
+    should be passed to the constructor of the ``DatasetReader``.
     """
+    def __init__(self, lazy: bool = False) -> None:
+        self._lazy = lazy
+
     def read(self, file_path: str) -> InstanceCollection:
         """
-        Actually reads some data from the `file_path` and returns a :class:`Dataset`.
+        Actually reads some data from the `file_path` and returns an :class:`InstanceCollection`.
+
+        The base class implementation passes this off to :func:`_read_instances`, and then uses the
+        resulting iterator to create either a :class:`~allennlp.data.dataset.Dataset` (which is
+        `not` lazy, and stores all of the instances in memory) or a
+        :class:`~allennlp.data.dataset.LazyDataset` (which `is` lazy, and reads, tokenizes, and
+        indexes the instances at every pass through the data).
+
+        If you want your ``DatasetReader`` to have this kind of lazy option built in, you should
+        override :func:`_read_instances`.  Otherwise, you can just override this method, and that
+        will still work fine with the API.
```

Comment:
I have mixed feelings about this guidance. Is there any reason that you shouldn't just implement `_read_instances`? (Other than that existing dataset readers don't do it?)

I'm all for giving people multiple ways to do things when there's a clear benefit to having multiple ways, but I don't see what that benefit is here.